Add ws support

hnaderi opened the pull request:
### What is done currently
- Implemented `WSClient` using the [connect-only approach](https://everything.curl.dev/libcurl/ws/concept) which is more aligned with `WSClient` interface
- Uses raw mode in order to receive pings
- websocket support in curl is experimental, so it needs to be built with `--enable-websockts`, which is added to the dev environment
- websocket client is not available when the host machine's curl does not support it, which is handled by optional constructor
- libcurl websocket implementation requires to know in advance, the total size of all fragments when using partial fragments, which can't be implemented due to `WSClient` interface, so this implementation does not support sending fragmented frames, however it can receive such frames.

### What needs to be done
- connect-only approach requires that its easy handle not be removed from multi-handle after connecting, which is not what happens currently
- using send and recv methods is causing `SIGSEGV` which is probably due to the former problem, which causes illegal memory access
- CI builds also need to be updated to have curl websockets available.

armanbilge commented on curl/src/main/scala/org/http4s/curl/WebSocketClient.scala:
```diff
@@ -0,0 +1,218 @@
+                val meta = stackalloc[libcurl.curl_ws_frame]()
+                var payload = ByteVector.empty
+
+                while ({
+                  throwOnError(
+                    libcurl.curl_easy_ws_recv(curl, buffer, bufSize, recv, meta)
+                  )
+
+                  payload = payload ++ ByteVector.fromPtr(buffer, (!recv).toLong)
+                  (!recv) == bufSize
+                }) {}
```
Is this blocking?

> * Implemented `WSClient` using the [connect-only approach](https://everything.curl.dev/libcurl/ws/concept) which is more aligned with `WSClient` interface

I'm curious why connect-only approach is a better fit? The callback based approach looks to be non-blocking.

hnaderi replied:
Yes, that's a blocking call unfortunately!
Connect-only approach eliminates the need to buffering again on receive side.

hnaderi replied:
I'll try callback or event based approach next, as doing it using connect only approach seems to be fruitless.

armanbilge replied:
Hmm, is it possible to pause/unpause recv to prevent buffering? This is how the HTTP client is working.
https://github.com/http4s/http4s-curl/blob/9354df6461146ff060f5d029c05b714f71ac7bd6/curl/src/main/scala/org/http4s/curl/CurlClient.scala#L128-L132

hnaderi replied:
I'm not sure whether that works with websocket transfer or not as websocket transfers seem to have a different lifecycle. Also websocket docs are very sparse and have some mismatches with published libcurl code too, I was and still am trying to build a working code to start experimenting with options available.

hnaderi replied:
I'll try that

armanbilge reviewed (commented):
Wow, nice!

hnaderi commented on flake.nix:
```diff
@@ -12,6 +12,13 @@
           inherit system;
           overlays = [ typelevel-nix.overlay ];
         };
+        curl = pkgs.curl.overrideAttrs (old: {
+          # Websocket support in curl is currently in experimental mode
+          # which needs to be enabled explicitly in build time in order to be available
+          #
+          #https://github.com/curl/curl/blob/f8da4f2f2d0451dc0a126ae3e5077b4527ccdc86/configure.ac#L174
+          configureFlags = old.configureFlags ++ [ "--enable-websockets" ];
+        });
```
Yes, I think that would be the right way to do it, however I'm not sure about the windows build.

armanbilge replied:
Yeah, I don't think we can do anything for Windows. Except disable the websocket test 😅

hnaderi commented:
I rewrote receive side from scratch, now it is asynchronous and most importantly, it works now!
However there are some problems to solve yet:
1. large fragments are not assembled correctly, I spent most of this time's effort to fix it, but I couldn't resolve the issue completely
2. This one is related to the ambiguities in the callback approach for websockets and docs,  I couldn't find how to distinguish an established connection from a connection that hasn't established yet; and it is important to know, because sending on a connection that is not established raises error (a segfault !!!)

I also checked pause/unpause docs, but I couldn't understand how it can help, as sends are blocking in websockets API and framing is done explicitly. Am I missing something?

armanbilge commented:
> I rewrote receive side from scratch, now it is asynchronous and most importantly, it works now!

Fantastic!!

---

> I also checked pause/unpause docs, but I couldn't understand how it can help, as sends are blocking in websockets API and framing is done explicitly. Am I missing something?

I mean, I don't know for sure, but I was thinking you can use it to prevent unbounded growth of incoming frames. Although the `Queue` is bounded, and unbounded amount of data may queue up waiting to be admitted to the queue. I _thought_ it's a general property of TCP connections, to be able to apply this sort of backpressure.

---

> 2\. I couldn't find how to distinguish an established connection from a connection that hasn't established yet

If configured, will curl invoke any read callbacks in this case? Maybe the first callback would indicate that the connection is established. (Using these callbacks may also help mitigate the blocking send). But the docs didn't mention whether read callbacks are are used for ws.

hnaderi commented:
> If configured, will curl invoke any read callbacks in this case? Maybe the first callback would indicate that the connection is established. (Using these callbacks may also help mitigate the blocking send). But the docs didn't mention whether read callbacks are are used for ws.

I tried read callbacks, transfer progress callbacks and header callbacks. The later works for this situation, first two are never called and in some configurations end in `CURLE_HTTP_RETURNED_ERROR (22)`.
Just fixed that using header callbacks.

> I mean, I don't know for sure, but I was thinking you can use it to prevent unbounded growth of incoming frames. Although the `Queue` is bounded, and unbounded amount of data may queue up waiting to be admitted to the queue. I _thought_ it's a general property of TCP connections, to be able to apply this sort of backpressure.

Hmm, Sounds reasonable :+1:  , I'll try implementing backpressure after fixing large payload issues.

hnaderi commented:
I've managed to fix the problem of large frames, it was in libcurl itself; libcurl `7.86.0` was the version that websocket support was introduced, but it had several framing issues, which are solved in `7.87.0`. I changed the build to include the correct version, and also disabled support for websockets on `7.86.0` completely, so if loaded libcurl in the env is not the right version, `WSClient` can't be constructed and also added related info in the runtime error.

By fixing that, it's working without problems and the only remaining feature is to implement back pressure using connection pause as you told.

hnaderi commented:
I just implemented a simple circuit breaker to apply back-pressure on websocket receive, and it works based on curl output; however I'm not sure and couldn't figure out whether or not pausing a websocket connection really stop buffering behind the scene too.

armanbilge commented:
> I just implemented a simple circuit breaker to apply back-pressure on websocket receive, and it works based on curl output

Cool, thanks for trying that! That's promising that it appears to work on the curl end. We can experiment with the ember websocket server to see if it works on the server-side. I expect it should!

Amazing work on this, it was not easy 😅 will take a closer look soon.

armanbilge commented on curl/src/main/scala/org/http4s/curl/websocket/Connection.scala:
```diff
@@ -0,0 +1,230 @@
+
+  def onEstablished(): Unit = dispatcher.unsafeRunAndForget(established.complete(()))
+
+  def onTerminated(result: Either[Throwable, Unit]): Unit = {
+    dispatcher.unsafeRunAndForget(receivedQ.offer(None))
+    result match {
+      case Left(err) =>
+        Console.err.println("Websocket connection terminated!")
+        err.printStackTrace(Console.err)
+      case _ => ()
+    }
+  }
```
I think if you do
```scala
dispatcher.unsafeRunAndForget(
  receivedQ.offer(None) *> IO.fromEither(result)
)
```

Then the error will be reported through the Cats Effect unhandled error mechanisms.

armanbilge commented on curl/src/main/scala/org/http4s/curl/websocket/CurlWSClient.scala:
```diff
@@ -0,0 +1,207 @@
+    val scheme = req.uri.scheme.getOrElse(ws)
+
+    if (scheme != ws && scheme != wss)
+      throw new RuntimeException(s"Websocket client can't handle ${scheme.value} scheme!")
```
```suggestion
      throw new IllegalArgumentException(s"Websocket client can't handle ${scheme.value} scheme!")
```

armanbilge commented on curl/src/main/scala/org/http4s/curl/websocket/CurlWSClient.scala:
```diff
@@ -0,0 +1,207 @@
+      )
+    }
+
+    var headers: Ptr[libcurl.curl_slist] = null
+    req.headers
+      .foreach { header =>
+        headers = libcurl.curl_slist_append(headers, toCString(header.toString))
+      }
+    throwOnError(
+      libcurl.curl_easy_setopt_httpheader(con.handler, libcurl_const.CURLOPT_HTTPHEADER, headers)
+    )
```
Are these being freed somewhere? The doc says this
> When this option is passed to [curl_easy_setopt](https://curl.se/libcurl/c/curl_easy_setopt.html), libcurl will not copy the entire list so you must keep it around until you no longer use this handle for a transfer before you call [curl_slist_free_all](https://curl.se/libcurl/c/curl_slist_free_all.html) on the list.

hnaderi replied:
Oops! No, I thought that current zone is handling that, but apparently those are raw pointer created outside.

hnaderi replied:
I open a new issue to resolve this one, as curl client has a similar issue and this pull request is way to huge by now!

armanbilge commented on curl/src/main/scala/org/http4s/curl/websocket/Error.scala:
```diff
@@ -0,0 +1,27 @@
+
+package org.http4s.curl.websocket
+
+sealed trait Error extends Throwable with Serializable with Product
```
Directly extending `Throwable` seems to be discouraged. Also since these are `object` should they extend `NoStackTrace` ?

https://contributors.scala-lang.org/t/nonfatal-and-controlthrowable-changed-in-stdlib/6049/5

```suggestion
sealed trait Error extends Exception with Serializable with Product
```

armanbilge commented on example/src/main/scala/WSExample.scala:
```diff
@@ -0,0 +1,55 @@
+
+}
+
+object WSEchoExample extends CurlApp.Simple {
+
+  private val echo = uri"wss://ws.postman-echo.com/raw"
+
+  def run: IO[Unit] = websocketOrError()
+    .connectHighLevel(WSRequest(echo))
+    .use { client =>
```
Can we adapt this one into a small test for the test suite? The test can be skipped if the curl is not properly configured.

hnaderi replied:
I'm working on tests currently, I'll update ASAP

hnaderi replied:
I've created a fake server and implemented tests against that server, however I have a problem which might be due to differences in scheduling; when I run websocket examples in ‍`CurlApp`, program runs and gracefully terminates as expected, but when I run it using `CatsEffectSuite` like the other existing suite, it runs and works but does not neither terminate nor timeout!  
I don't have any idea what could be the source of problem! have you encountered this problem before? @armanbilge

armanbilge replied:
Yes I have, it probably means that the number of running handles has not returned to `0`, so it keeps looping. Is the handle used for the WS connection being deregistered/canceled somewhere?

https://github.com/http4s/http4s-curl/blob/ea998518ebea920098811579c979b161a03c8bca/curl/src/main/scala/org/http4s/curl/unsafe/CurlExecutorScheduler.scala#L82

hnaderi replied:
I guess it does, I added logs on finalization and it seems to work, and above all it works as expected when runs under `CurlApp` but not under `CatsEffectSuite`

hnaderi replied:
Just pushed it

hnaderi replied:
No, it should work fine.

armanbilge replied:
Yeah I'm really confused what's happening.

```
$ uname -a
Linux http4s-http4scurl-pzxb58choln 5.15.0-47-generic #51-Ubuntu SMP Thu Aug 11 07:51:15 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

$ nix --version
nix (Nix) 2.11.0
```

armanbilge replied:
No, my shell seems fine. It's a toolchain problem not picking up the right lib for some reason.
```
[http4s-curl]$ nm $LIBRARY_PATH/libcurl.so | grep curl_ws_meta
0000000000071d20 T curl_ws_meta
```

armanbilge replied:
Lol, so I added this.
```scala
  } else {
    c.withLinkingOptions(
      s"-L${System.getenv("LIBRARY_PATH")}" +: c.linkingOptions
    )
```
and now I moved on to this 😂 
```
[error] /nix/store/178vvank67pg2ckr5ic5gmdkm3ri72f3-binutils-2.39/bin/ld: /nix/store/9m94s7xxwqa57yj33maxsv03iqa009m3-devshell-dir/lib/libcurl.so: undefined reference to `nghttp2_option_set_no_rfc9113_leading_and_trailing_ws_validation'
```

hnaderi replied:
I did a lot of tracing and I'm pretty confident that `curl_easy_cleanup` is called. But it seems that websocket transfers don't change their status on call to `curl_multi_info_read`
So it's not detected for removal from `multi handle` in the scheduler.
However sending a close frame explicitly causes it to be cleaned up, which is not a solution I guess.

armanbilge replied:
> But it seems that websocket transfers don't change their status on call to `curl_multi_info_read`

Hmm, that seems like it can be an issue, since that's also used for cleaning up the stored callback.



> However sending a close frame explicitly causes it to be cleaned up, which is not a solution I guess.

Why not? Do you mean server or client sending a close frame? I assume the client always wants to send a close frame, no?

hnaderi replied:
I think adding a new `CurlExecutorScheduler.addHandler` that returns a `Resource` can solve the problem.

hnaderi replied:
> Why not? Do you mean server or client sending a close frame? I assume the client always wants to send a close frame, no?

I mean the Client, but if it depends on the server to respond it would useless, as interrupted connections won't ever terminate.

hnaderi replied:
I added a few tests as you suggested, however it is causing the build to fail currently.

armanbilge reviewed (commented):
Looks great, took a passthrough with some nits and questions.

hnaderi commented on build.sbt:
```diff
@@ -44,12 +44,14 @@ ThisBuild / nativeConfig ~= { c =>
       .withLinkingOptions(c.linkingOptions :+ s"-L${vcpkgBaseDir}/installed/x64-windows/lib/")
   } else c
+
+  platformOptions.withLinkingOptions(platformOptions.linkingOptions :+ "-lcurl")
```
Yes, and it does; however due to ordering important in linker params, there were some problems compiling the c file.
Without this option, linking would fail when want to link `curl.c`.
It happens because I added a few function calls in the c file (for version info), which need to be linked too.

hnaderi commented:
I finally managed to fix the problem after more tracing and reading lib curl code.
The problem, as I mentioned yesterday, was that websocket transfers behave differently;
so adding such a handle to `CurlExecutorScheduler` like other handlers is wrong, as their status won't change, so
```scala
          val info = libcurl.curl_multi_info_read(multiHandle, msgsInQueue)
```
will always be null, and its callback is neither called nor removed, even if you remove the handle from the multi handle;
which causes application to hang when everything is done, because cats effect quere is empty and has nothing to do and calls the `poll` method with infinite timeout, so scheduler calls `curl_multi_poll` which never returns as curl also does not have any other task in its queue.

hnaderi commented:
I was thinking of separating websocket to a different sbt project, so that users who don't care about websocket stay away from experimental stuff and its headaches; it might also help with builds.  
What's your opinion?

armanbilge commented:
> I finally managed to fix the problem after more tracing and reading lib curl code.

The fix looks good! 👍 

---

> I was thinking of moving websocket to a separate sbt project, so that users who don't care about websocket stay away from experimental stuff and its headaches;

Hmm yes that does sound attractive. On the other hand I think since Scala Native does link-time dead code elimination, as long as a user doesn't touch the websocket stuff, then they will not need a curl with websockets enabled. So it's mostly about deciding what works best for build/CI.

armanbilge commented on build.sbt:
```diff
@@ -83,3 +67,56 @@ lazy val example = project
     )
   )
+
+lazy val testServer = project
+  .in(file("test-server"))
+  .enablePlugins(NoPublishPlugin)
+  .settings(
+    libraryDependencies ++= Seq(
+      "org.typelevel" %% "cats-effect" % catsEffectVersion,
+      "org.http4s" %% "http4s-dsl" % http4sVersion,
+      "org.http4s" %% "http4s-ember-server" % http4sVersion,
+      "ch.qos.logback" % "logback-classic" % "1.2.6",
+    )
+  )
```
I also use a test server in http4s-dom, but I run it as part of the meta-build.

https://github.com/http4s/http4s-dom/blob/51c42d18f0d5a7d8ebc79bb47e0960bb4bfc6b35/build.sbt#L47-L48

It's not perfect but it works.

hnaderi replied:
I'm not sure if I got your point, could you please elaborate?

hnaderi commented:
Finally!!!
After a few days of fighting with a lot of annoying build issues in CI, I managed to add websocket tests which required having a custom built curl.  
I tried several options, and then tried to narrow them to the simplest ones and the result is the one with least magic, which IMHO should be used as a doc for people who want to use websocket client as well.

I consider this to be done, please review the new changes @armanbilge

armanbilge commented on curl/src/main/scala/org/http4s/curl/unsafe/libcurl.scala:
```diff
@@ -59,6 +61,18 @@ private[curl] object libcurl_const {
   final val CURL_READFUNC_PAUSE = 0x10000001L
+
+  // constant flags from websocket.h
+  final val CURLWS_TEXT = 1 << 0
+  final val CURLWS_BINARY: CInt = 1 << 1
+  final val CURLWS_CONT: CInt = 1 << 2
+  final val CURLWS_CLOSE: CInt = 1 << 3
+  final val CURLWS_PING: CInt = 1 << 4
+  final val CURLWS_OFFSET: CInt = 1 << 5
+  final val CURLWS_PONG: CInt = 1 << 6
```
```suggestion
  final val CURLWS_TEXT = 1 << 0
  final val CURLWS_BINARY = 1 << 1
  final val CURLWS_CONT = 1 << 2
  final val CURLWS_CLOSE = 1 << 3
  final val CURLWS_PING = 1 << 4
  final val CURLWS_OFFSET = 1 << 5
  final val CURLWS_PONG = 1 << 6
```

armanbilge reviewed (approved):
That was a fierce battle! I like how you setup the build / CI 👍

I don't have any new comments and I see the WS suite is running in CI, so LGTM ✅ thank you for all of your work on this! Suffice to say, it was even harder than I thought it would be 😬 

I guess we should probably track down the slist memory leak, and then do you want to make a release? 😁

hnaderi commented:
> I guess we should probably track down the slist memory leak, and then do you want to make a release? 😁 

Yes, I work on that in a few days, and also I think I can also handle a few other issues before having a release.
